### ISSUE TITLE
Support test matrix

### ISSUE THREAD

**Issue by electrum** [junit-team/junit5#710]
Suppose we're writing an HTTP client library that is a wrapper around several different backend libraries. We support many options and we need to test all the combinations.

Protocol and security:

* HTTP over HTTP/1
* HTTPS over HTTP/1
* HTTP over HTTP/2
* HTTPS over HTTP/2

Backend library:

* Apache HTTP Client
* Jetty HTTP Client
* OkHttp

SOCKS proxy:

* disabled
* enabled

We can write a test class `AbstractClientTests` that has methods like `testPostMethod`, `testReadTimeout`, `testResponseStatus`, etc. To run the class for each combination of the above, we create a subclass for each combination.

This approach has several advantages:

* `AbstractClientTests` is readable and looks like a normal test class
* we can run and debug a specific test in the IDE
* tools like Maven Surefire understand this structure, since everything is a normal test class

Of course, there are big disadvantages:

* we need `4 x 3 x 2 = 24` subclasses
* we can't split up `AbstractClientTests`, since duplicating all the subclasses would be too painful

This doesn't seem like a good use case for _dynamic tests_, which requires turning everything inside out and losing tool support. It seems what we need is _dynamic test classes_.

**Comment by sormuras** [comment]
[TestTemplate](https://github.com/junit-team/junit5/pull/642) and [Params](https://github.com/junit-team/junit5/issues/14) to the rescue! There's **no** explicit `MatrixTest` support, but it can easily achieved with these frameworks.

I'm just in the process of squashing a test matrix into a flattened dynamic test. It works, but it doesn't feel awesome. Here is the 3-times nested for-loop generating the dynamic test instances https://github.com/junit-team/junit5/compare/refactor_tree_printing_listener#diff-bcdd220ee20a4cdae2c4835d9e21e65bR150

**Comment by marcphilipp** [comment]
We're only days (I hope) away of having first-class support for parameterized tests. Using those, you could write your tests like this:

```java
@ParameterizedTest
@MethodSource(names = "allCombinations")
void testPostMethod(Protocol protocol, BackendLibrary backend, SocksProxy socksProxy) {
    // test something
}

@ParameterizedTest
@MethodSource(names = "allCombinations")
void testReadTimeout(Protocol protocol, BackendLibrary backend, SocksProxy socksProxy) {
    // test something
}

@ParameterizedTest
@MethodSource(names = "allCombinations")
void testResponseStatus(Protocol protocol, BackendLibrary backend, SocksProxy socksProxy) {
    // test something
}

static List<Arguments> allCombinations() {
    // compute cross product of three sets somehow or define it manually:
    return asList(
        ObjectArrayArguments.create(Protocol.HTTP1, BackendLibrary.APACHE, SocksProxy.ENABLED),
        ObjectArrayArguments.create(Protocol.HTTP1, BackendLibrary.APACHE, SocksProxy.DISABLED),
        ...
    );
}
```

If that's too involved you could alternatively create your own [`TestTemplateInvocationContextProvider`](https://github.com/junit-team/junit5/blob/master/junit-jupiter-api/src/main/java/org/junit/jupiter/api/extension/TestTemplateInvocationContextProvider.java) that generates the cross-product for you and use it like this:

```java
@ExtendWith(MyCrossProductInvocationContextProvider.class)
class ClientTests {

    @TestTemplate
    void testPostMethod(Protocol protocol, BackendLibrary backend, SocksProxy socksProxy) {
        // test something
    }

    @TestTemplate
    void testReadTimeout(Protocol protocol, BackendLibrary backend, SocksProxy socksProxy) {
        // test something
    }

    @TestTemplate
    void testResponseStatus(Protocol protocol, BackendLibrary backend, SocksProxy socksProxy) {
        // test something
    }
}
```

**Comment by electrum** [comment]
Thanks for the detailed responses!

The `@ParameterizedTest` version is rather verbose for this use case. It looks very similar to `@Test(dataProvider = "allCombinations")` and `@DataProvider(name = "allCombinations")` in TestNG. But it's definitely a handy feature for other cases that we use today in TestNG.

The `TestTemplateInvocationContextProvider` approach looks much better. For these types of use cases, the parameters are normally provided to a factory, and the tests just get the final object to test:

```java
@ExtendWith(HttpClientTestContextProvider.class)
class ClientTests {

    @TestTemplate
    void testPostMethod(HttpClient client) {
        // test something
    }

    @TestTemplate
    void testReadTimeout(HttpClient client) {
        // test something
    }
}
```

How does lifecycle and parallelism work with this? There are a few additional concerns that I forgot about earlier:

* The objects to test can be expensive to create, so we don't want to create them for every test method. For example, an HTTP client might require setting up SSL contexts, thread pools, etc.
* The objects might be so expensive that we want to share a single instance across multiple test classes.
* The objects need to be destroyed when no longer used in order to free up resources.
* The objects to test might not be thread safe, so if we are running test methods in parallel, we need to create a new instance per test thread.

With TestNG, we use the "methods" parallel mode, which runs each method in a separate thread. However, TestNG only ever creates a single instance of the test class, so that requires either the test object to be thread safe, or to create the object within each test method. In practice, most expensive test objects are thread safe, so having a single one at the class level suffices.

**Comment by electrum** [comment]
Side note: building a test matrix generically (when you can't hard-code the nested for-loops) can be done easily with Guava's [Sets.cartesianProduct()](https://google.github.io/guava/releases/snapshot/api/docs/com/google/common/collect/Sets.html#cartesianProduct-java.util.List-)

**Comment by sbrannen** [comment]
@sormuras, the link you posted is broken. Do you have an updated link?

**Comment by sbrannen** [comment]
> * The objects to test can be expensive to create, so we don't want to create them for every test method. For example, an HTTP client might require setting up SSL contexts, thread pools, etc.
> * The objects might be so expensive that we want to share a single instance across multiple test classes.

For those needs you can store things in the `ExtensionContext.Store` between invocations.

> * The objects need to be destroyed when no longer used in order to free up resources.

In theory, if you store things in the `ExtensionContext.Store` they should get garbage collected when the enclosing scope (e.g., a test class) has completed its execution. I say "in theory" because that is the intended behavior, but I don't believe it has been thoroughly tested.

> * The objects to test might not be thread safe, so if we are running test methods in parallel, we need to create a new instance per test thread.

Unlike TestNG, JUnit (versions 1-4 and currently Jupiter as well) instantiates a new instance of the test class for every test method (or test template) execution. So state in the test instance should not be a problem.

Note, however, that JUnit Jupiter does not currently have built-in support for parallel test execution, but that might come in M5.

**Comment by sormuras** [comment]
Sure: https://github.com/junit-team/junit5/blob/master/platform-tests/src/test/java/org/junit/platform/console/ConsoleDetailsTests.java#L149

**Comment by sbrannen** [comment]
Aha.... definitely makes me feel like _I'm in the Matrix_. 😝

**Comment by electrum** [comment]
> In theory, if you store things in the ExtensionContext.Store they should get garbage collected when the enclosing scope (e.g., a test class) has completed its execution.

Relying on garbage collection to free resources is not reliable. We can run out of threads, file descriptors, etc., long before the garbage collector kicks in. Having proper lifecycle of the objects (e.g., implement `Closeable`) is required.

> Unlike TestNG, JUnit (versions 1-4 and currently Jupiter as well) instantiates a new instance of the test class for every test method

Great! This is a huge pain point with parallel tests in TestNG.

**Comment by sbrannen** [comment]
> Relying on garbage collection to free resources is not reliable. We can run out of threads, file descriptors, etc., long before the garbage collector kicks in. Having proper lifecycle of the objects (e.g., implement `Closeable`) is required.

That's a great point... and something we have not yet considered. I'll add a new issue to track that.

Cheers!

**Comment by sbrannen** [comment]
Done: see #742.

**Comment by marcphilipp** [comment]
> Relying on garbage collection to free resources is not reliable. We can run out of threads, file descriptors, etc., long before the garbage collector kicks in. Having proper lifecycle of the objects (e.g., implement `Closeable`) is required.

This is now possible: 5.1 introduced `Store.CloseableResource`.

In addition to parameterized methods, #878 will introduce parameterized classes which should cut down the boilerplate for the scenario described in this issue. Thus, I'm closing it for now. Please let us know if you have any questions.
